### DISCUSSION TITLE
Could you help me understand memory barriers?

### DISCUSSION THREAD

**Discussion by shi-yan** [gpuweb/gpuweb#4401] [Q&A] [answered]
I can't understand storageBarrier, texture Barrier and workgroupBarrier,

for example, in CUDA, we have __syncthreads(), which has nothing to do with memory. all threads simply stop here before continuing. This is understandable. 

But what does a `workgroupBarrier` mean? if I don't access a workgroup allocation at all, will the synchronization work at all?

for example:

```
A thread {
...
var a:i32 = 1;
for (var i:i32=0;i<100;++i)
{
a = a + 1;
workgroupBarrier();  // do we have thread synchronization here at all? what if I use storageBarrier or textureBarrier here? any differences?
}
...
}
```

Thanks.

**Comment by greggman** [comment, marked as answer]
There is this: https://stackoverflow.com/questions/72035548/what-does-storagebarrier-in-webgpu-actually-do

copied from that answer

> * storageBarrier coordinates access by invocations in single workgroup to buffers in 'storage' address space.
>
> * workgroupBarrier coordinates access by invocations in a single workgroup to variables in 'workgroup' address space.
>
> In detail, a reasonable way to think about it is that a barrier for address space X (X is 'workgroup' or 'storage), is a point in execution where:
>
> * all invocations in a workgroup wait for each other to reach the barrier
>
> * all in-flight writes to variables in address space 'X' complete
>
> * then all invocations become unblocked, and can continue executing after the barrier.
>
> * after the barrier, any reads from variables in address space 'X' will "see" the writes that were initiated before the barrier.

see the full answer for more details

**Reply by shi-yan** [reply]
Thank you! Yes, I found this link before asking. The part I don't understand is if I have no workgroup allocation access, what does `workgroupBarrier` do?

**Reply by greggman** [reply]
I'm only guessing from the description that it does what it says above, minus the irrelevant parts

> all invocations in a workgroup wait for each other to reach the barrier

Yes

> all in-flight writes to variables in address space 'X' complete

If there are no writes, this is irrelevant

> then all invocations become unblocked, and can continue executing after the barrier.

Yes

> after the barrier, any reads from variables in address space 'X' will "see" the writes that were initiated before the barrier.

If there are no reads, this is irrelevant.

**Reply by dneto0** [reply]
Yes, @greggman is right.

When there are no variables shared between the invocations, then the only effect of the barrier is to make all the invocations in the workgroup wait for each other before proceeding.

The code sample you have doesn't have any shared variables.  It would be weird to use the barrier in this case. That's because the only way one invocation can influence another is by not getting to the barrier at all. But that will hang the compute shader and then the OS will likely intervene and kill the process. You probably don't want to do that.


>  in CUDA, we have __syncthreads(), which has nothing to do with memory.

That's not right.

https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html?highlight=__syncthreads#synchronization-functions says

> void __syncthreads();
> waits until all threads in the thread block have reached this point and all global and shared memory accesses made by these threads prior to __syncthreads() are visible to all threads in the block.

Translation: CUDA global memory = WGSL storage buffers;  CUDA shared memory = WGSL workgroup memory.

So CUDA __syncthreads() does both WGSL workgroupBarrier() *and* storageBarrier() combined.
